library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java

```diff
@@ -0,0 +1,516 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.base.api.util;
+
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * Represents a monad that can either contain a value or no value.<br/>
+ * Similar to the {@link Optional} class but with some performance improvements
+ * and quality of life changes.<br/>
+ *
+ * Can either be {@link Just} or {@link Nothing}.<br/>
+ * {@link Just} instances contain a value of type {@link T}.<br/>
+ * {@link Nothing} instance all reference the {@link Nothing} singleton and contain no data.<br/>
+ *
+ * @param <T> The type of the wrapped object
+ * @author 0xJoeMama
+ */
+public sealed abstract class Maybe<T> permits Maybe.Just, Maybe.Nothing {
+	/**
+	 * Similar function to {@link Optional#ofNullable} function.<br/>
+	 * Wraps a {@link T} value into an {@link Maybe}.
+	 *
+	 * @param value A {@link Nullable} value that will be wrapped into a {@link Maybe}.
+	 * @param <T>   The type of object the {@link Maybe} instance will hold.
+	 * @return An instance of {@link Just} if the provided value is not <code>null</code>,<br/>
+	 * otherwise the singleton {@link Nothing} instance.
+	 * @see Optional#ofNullable
+	 */
+	public static <T> Maybe<T> wrap(@Nullable T value) {
+		return value != null ? just(value) : nothing();
+	}
+
+	/**
+	 * Similar function to {@link Optional#of}.<br/>
+	 * Wraps the provided value into a {@link Just} instance.
+	 *
+	 * @param value A {@linkplain org.jetbrains.annotations.NotNull non-null} value that is to be wrapped.
+	 * @param <T>   The type of object the {@link Maybe} instance will hold.
+	 * @return A {@link Just} instance containing the provided value.
+	 * @see Optional#of
+	 */
+	public static <T> Maybe<T> just(T value) {
+		return new Just<>(value);
+	}
+
+	/**
+	 * Similar function to {@link Optional#empty()}.
+	 *
+	 * @param <T> The type the {@link Nothing} instance will take the form of.
+	 * @return the singleton {@link Nothing} instance after casting it to the proper type.
```

Comment:
```suggestion
	 * @return A {@link Nothing} instance.
```
Exposing this implementation details is just gonna make people do `maybe == Maybe.nothing()`.